### PR TITLE
*: show more region info in api

### PR DESCRIPTION
**Pull request by nolouch**
close https://github.com/pingcap/pd/issues/583
PTAL @disksing @siddontang @andelf

### CODE REVIEW

**Comment by siddontang** [comment]
Please fix the conflict first.

File: pdctl/command/region_command.go [hunk 1]
```diff
@@ -101,28 +100,14 @@ func showRegionWithTableCommandFunc(cmd *cobra.Command, args []string) {
...
 		return
 	}
-
-	client, err := getClient()
-	if err != nil {
-		fmt.Println("Error: ", err)
-		return
-	}
-	region, leader, err := client.GetRegion(context.Background(), key)
+	prefix := fmt.Sprintf(regionKeyPrefix, key)
```
Comment (disksing) [inline pdctl/command/region_command.go]: Any needs of using `url.PathEscape()`?

File: pdctl/command/region_command.go [hunk 2]
```diff
@@ -15,20 +15,19 @@ package command
...
 	"github.com/spf13/cobra"
-	"golang.org/x/net/context"
 )
 
 var (
-	regionsPrefix = "pd/api/v1/regions"
-	regionPrefix  = "pd/api/v1/region/%s"
+	regionsPrefix   = "pd/api/v1/regions"
+	regionIDPrefix  = "pd/api/v1/region/id/%s"
+	regionKeyPrefix = "pd/api/v1/region/key/%s"
```
Comment (siddontang) [inline pdctl/command/region_command.go]: prefix should not contain `%s` here, you can format whole key explicitly.

Comment (siddontang) [inline pdctl/command/region_command.go]: or use string add directly

File: server/cluster.go [hunk 3]
```diff
@@ -291,14 +291,24 @@ func (s *Server) bootstrapCluster(req *pdpb.BootstrapRequest) (*pdpb.BootstrapRe
...
 	if region == nil {
 		return nil, nil
 	}
 	return region.Region, region.Leader
 }
 
+// GetRegionInfoByKey gets regionInfo by region key from cluster.
+func (c *RaftCluster) GetRegionInfoByKey(regionKey []byte) *RegionInfo {
+	region := c.cachedCluster.searchRegion(regionKey)
+	if region == nil {
```
Comment (siddontang) [inline server/cluster.go]: you can return region here directly, no need checking nil here

File: server/cluster.go [hunk 4]
```diff
@@ -308,6 +318,15 @@ func (c *RaftCluster) GetRegionByID(regionID uint64) (*metapb.Region, *metapb.Pe
 	return region.Region, region.Leader
 }
 
+// GetRegionInfoByID gets regionInfo by regionID from cluster.
+func (c *RaftCluster) GetRegionInfoByID(regionID uint64) *RegionInfo {
+	region := c.cachedCluster.getRegion(regionID)
+	if region == nil {
```
Comment (siddontang) [inline server/cluster.go]: ditto

**Comment by siddontang** [comment]
what does the pd-ctl output now?

File: pdctl/command/region_command.go (hunk 1 above)
Comment (nolouch) [inline pdctl/command/region_command.go]: the function `PathEscape` is new add in go 1.8. It is enough to deal with the key firstly with the protobuf format.

**Comment by nolouch** [comment]
like
```
{
  "id": 2,
  "end_key": "bURCOjMAAAD/APsAAAAAAAD/AGhUYWJsZTr/NwD+AAAAAAD6",
  "region_epoch": {
    "conf_ver": 9,
    "version": 2
  },
  "peers": [
    {
      "id": 44,
      "store_id": 7
    },
    {
      "id": 65,
      "store_id": 1
    },
    {
      "id": 17259,
      "store_id": 6
    }
  ],
  "Leader": null,
  "DownPeers": [],
  "PendingPeers": []
}

```

**Comment by siddontang** [comment]
I think we should support protobuf string output format for start and end key later. 

LGTM

**Comment by disksing** [comment]
LGTM

**Pull request merged**